### ISSUE TITLE
Add JSONSchema typings to AJV

### ISSUE THREAD

**Issue by junalmeida** [ajv-validator/ajv#972]
**What version of Ajv you are you using?**
v6

**What problem do you want to solve?**
Use a type in typescript for json schema parameters.

**What do you think is the correct solution to problem?**
Use this typings
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/json-schema/

**Will you be able to implement it?**
No

**Comment by rkma** [comment]
use javascript - ajv validaion - if i give a json as a string format - is AJV accepts,
And where parse json object to ajv if error in required condition ....is able to  display which line error occured???

**Comment by Relequestual** [comment]
What exactly is being asked here?
This library is not written in TS.

**Comment by epoberezkin** [comment]
I am not quite sure what is the request - is it about compiling JSON Schema to typescript types? Then it’s definitely out of scope.

**Comment by junalmeida** [comment]
A library not written in TS is not a impediment to have TS declarations to be used in a TS project.

**Comment by Relequestual** [comment]
Regardless, as someone who doesn't use TS, I cannot understand what you are asking for or why you want it.

Your answer to "What problem do you want to solve?" is not enough. It just says "DO X". 🤷‍♂️

**Comment by epoberezkin** [comment]
@junalmeida there are TS typings for the function parameters. Are you suggesting to add type declaration to parameters that are schemas? As opposed to using plain object?

**Comment by epoberezkin** [comment]
It does does seem so looking at the link.

There are two problems with that approach:
1. It’s redundant as the schemas are validated at run time (only once) and you can do it as part of the test suite.
2. There is no way to comprehensively map validation against meta schema to TS types.

The second part is more important - I’ve seen different approach and given that schema has a wider set of possible restrictions than TS types, you may choose to compile TS typings for your data to schemas, or to compile schemas to types, etc. All approaches have merits.

Closing the issue.

**Comment by junalmeida** [comment]
I am suggesting to use the correct type for JSON schema (I put the link) instead of `any`

**Comment by epoberezkin** [comment]
I don’t think it would be the correct approach - you cannot and shouldn’t ascertain the validity of JSON Schema itself using typescript types - I provided the reasons above. The types in the link you sent are not sufficient and not necessary to ascertain validity.

Also it is ok to have additional / custom keywords in the schema and types would prevent it. Ajv has strict mode that flags any unknown keywords.

**Comment by junalmeida** [comment]
Sorry but I don't agree with either points: 

> It’s redundant as the schemas are validated at run time (only once) and you can do it as part of the test suite.

So, the purpose of typescript is not to validate anything at run time, but to provide a consistent type checking at design/compile time. So, it is not redundant, it is a guide during the time you are writing your code, and a valuable help on refactoring if you need.

> There is no way to comprehensively map validation against meta schema to TS types.

I didn't get why. I used this TS typings casting from your `any` type to the correct JSON schema type with no problem.

**Comment by epoberezkin** [comment]
> I didn't get why. I used this TS typings casting from your any type to the correct JSON schema type with no problem.

TS typings are less restrictive than the meta-schema.

You can add typings to yhe app from the link you posted.

Also, it will not work with custom keywords - that alone is enough to not use it.

**Comment by junalmeida** [comment]
What do you mean by custom keywords? I'm using my custom schema, with custom keywords without any problem. I just had to extend the JSONSchema because the `errorMessage` property is not part of the official JSONSchema.org. 

This is the extension I implemented myself:

```typescript

export interface AjvJSONSchema {
    errorMessage?: string | AjvSchemaErrorMessage;
}

interface AjvSchemaErrorMessage {
    type?: string;
    required?: string | AjvSchemaErrorMessageDefinition;
    properties?: string | AjvSchemaErrorMessageDefinition;
    additionalProperties?: string;

}

interface AjvSchemaErrorMessageDefinition {
    [key: string]: string;
}
```

And this is my custom validation schema as an example to be validated at runtime:
```typescript
    private validationSchema: JSONSchema7 & AjvJSONSchema = {
        type: "object",
        required: [
            "country",
            "currency",
            "provider",
            "accountNumber",
        ],
        properties: {
            country: { type: "string", minLength: 1, maxLength: 2 },
            currency: { type: "string", minLength: 1, maxLength: 3 },
            provider: { type: "string", minLength: 1, maxLength: 10 },
            accountNumber: { type: "string", minLength: 1, maxLength: 30 },
            routingNumber: { type: "string", minLength: 1, maxLength: 30 },
        },
        errorMessage: {
            required: {
                country: "Please select a country",
                currency: "Please select a currency",
                provider: "Please select a payment provider",
                accountNumber: "Please type in an account number.",
            },
        },
    };
```

So, my suggestion is to change `ajv.d.ts` on line 42 to the following: 
```typescript
    validate(schemaKeyRef: JSONSchema7 & AjvJSONSchema | string | boolean, data: any): boolean | PromiseLike<any>;
```
Same for `compile`, `compileAsync` and so on...

**Comment by briancodes** [comment]
@junalmeida Those typings only work for the top level properties, but won't work if `errorMessage` is used for more fine grain messages e.g. 
```
{
  ...
  properties: {
    country: { 
      type: "string", 
      minLength: 1, 
      maxLength: 2 ,
      errorMessage: {
        minLength: "Country should be greater than one"
      }
    }
  }
}
```

**Comment by junalmeida** [comment]
@briancodes Yes, that's why I've added a custom extension to the base `JSON7Schema` as you can see in my code. My suggestion is to use `JSON7Schema` and add the `errorMessage` extension to typings of `AJV`.

**Comment by briancodes** [comment]
@junalmeida  I'm saying that the custom extension you used `JSONSchema7 & AjvJSONSchema` doesn't work for anything other than the top level item

**Comment by briancodes** [comment]
@junalmeida I think the custom extension you described, `JSONSchema7 & AjvJSONSchema` will only work if `errorMessage` is a top level property, but won't work for many of the other use cases of `errorMessage` e.g. if it's added to properties at a deeper level in the schema

**Comment by junalmeida** [comment]
@briancodes Ah, yes, I see your point now! My code was just an idea, we can certainly improve the typing to achieve multiple levels where applicable.
